cupy/core/fusion.py

```diff
@@ -239,99 +344,192 @@ def copy(self):
+
+    def _fresh_postmap_local(self, *args, **kwargs):
+        index = self.fresh_index()
+        var = _FusionVarCUDA(index, *args, **kwargs)
+        self.postmap_local_list.append(var)
+        return var
+
+    def fresh_local(self, *args, **kwargs):
+        if self.has_reduction():
+            return self._fresh_postmap_local(*args, **kwargs)
+        else:
+            return self._fresh_premap_local(*args, **kwargs)
+
+    def _add_premap_op(self, *args, **kwargs):
+        op = FusionOp(len(self.op_list), *args, **kwargs)
+        subm = op.submodule
+        self.submodules[subm.key()] = subm
+        self.op_list.append(op)
+        self.add_preamble(subm.preamble)
+        return op
+
+    def _add_postmap_op(self, *args, **kwargs):
+        op = FusionOp(len(self.postmap_op_list), *args, **kwargs)
+        subm = op.submodule
+        self.submodules[subm.key()] = subm
+        self.postmap_op_list.append(op)
+        self.add_preamble(subm.preamble)
+        return op
+
+    def add_op(self, *args, **kwargs):
+        if self.has_reduction():
+            return self._add_postmap_op(*args, **kwargs)
+        else:
+            return self._add_premap_op(*args, **kwargs)
+
+    def set_reduce_op(self, raw, arg, kwargs):
+        for op in raw._ops:
+            (input_type,), (output_type,), _ = op
+            if numpy.can_cast(arg.dtype.type, input_type):
+                return_dtype = numpy.dtype(output_type)
+                self.premap_ret = self.get_cuda_var(arg)
+                self.reduce_op = op
+                self.reduce_identity = raw.identity
+                self.reduce_kwargs = kwargs
+                self.add_preamble(raw._preamble)
+                return self.fresh_postmap_param(return_dtype)
+        return None
+
+    def add_preamble(self, preamble):
+        self.preamble_set.add(preamble)
+
+    def get_cuda_var(self, arg):
+        """This converts `arg` to _FusionVarCUDA data.
+
+        Args:
+            arg (FusionVarPython or a primitive type)
+
+        Return value: _FusionVarCUDA
+        """
+        arg_type = type(arg)
+        if arg_type is FusionVarPython:
+            if arg._is_postmap == self.has_reduction():
+                return arg._var
             else:
-                if not numpy.can_cast(args[i].const, ty_ins[i]):
-                    return False
-        return True
+                # Map operation between pre-map variable and post-map variable
+                raise Exception('Shape mismatch')
+        is_scalar = arg_type in six.integer_types + (float, bool, complex)
+        is_ndarray = hasattr(arg, 'dtype') and arg.dtype in _dtype_list
+        if is_scalar or is_ndarray:
+            return self.fresh_local(numpy.dtype(arg_type), const=arg)
+        raise Exception('Unsupported type {}'.format(arg_type))
+
+    def call_ufunc(self, ufunc, args, kwargs):
+        nin = ufunc.nin
+        nout = ufunc.nout
+
+        def _should_use_min_scalar(in_args):
```

Comment:
Please write the brief description of `_should_use_min_scalar`, `can_cast1` and `can_cast2`, respectively.

Reply:
Added a brief note.